### PR TITLE
context : perform output reorder lazily upon access after sync

### PR DESCRIPTION
**Pull request by ggerganov**
ref https://github.com/ggml-org/llama.cpp/issues/14795#issuecomment-3113050944

After processing a batch, remember the indices that we have to swap and apply the data swap (of logits and embeddings) upon access via `llama_get_...`

### CODE REVIEW

**Comment by slaren** [comment]
Can we avoid the reorder entirely if we drop `llama_get_logits` in favor of `llama_get_logits_ith`?

**Comment by ggerganov** [comment]
Hm, I think `llama_get_logits_ith` also requires the reorder?

**Comment by ggerganov** [comment]
Ah, got it. Let me see.

**Comment by ggerganov** [comment]
Yes, we can "redirect" the index on-the-fly using the `output_ids`. Seems like a good idea.

Do it in a follow-up PR, so we can patch the current issue for now?

**Review by slaren — APPROVED** [review]

**Comment by compilade** [comment]
> Hm, I think `llama_get_logits_ith` also requires the reorder?

Only because the `out_ids` are sorted, otherwise it wouldn't.

https://github.com/ggml-org/llama.cpp/blob/e4868d16d24dec55e61bcaadaca28feed8f98b13/src/llama-context.cpp#L1221-L1225

If `llama_get_logits` and `llama_get_embeddings` are removed, then the sorting and output reordering can be removed altogether.

(Right, this comment is redundant with <https://github.com/ggml-org/llama.cpp/pull/14853#issuecomment-3113143522> and <https://github.com/ggml-org/llama.cpp/pull/14853#issuecomment-3113460013>)

**Comment by compilade** [comment]
Actually, because of how `llama_get_logits_ith` is called in `llama-perplexity`, sorting is still required.

https://github.com/ggml-org/llama.cpp/blob/e4868d16d24dec55e61bcaadaca28feed8f98b13/tools/perplexity/perplexity.cpp#L612

The above assumes the logits after the ith one are in the correct order.

There might need to be an API for logits ranges.

**Comment by ggerganov** [comment]
This usage of `llama_get_logits_ith()` is technically wrong because it assumes the logits are sequential, but this is not stated in the API. So I think it is OK to break it.

**Comment by slaren** [comment]
It is indirectly stated in the comment that says `Equivalent to: llama_get_logits(ctx) + ctx->output_ids[i]*n_vocab`. The examples would need to be updated, but that should be an easy problem to solve, the bigger problem is that it may be break silently 3rd party code that we have no control over.
